In [ ]:
#### Exercise

prompt = "   Whoa, this text is awesome. python is a language for those who think big. Who else thinks python is awesome? python is wild.   "

print(prompt)

print(prompt.lstrip())

print(prompt.rstrip())

clean_prompt = prompt.strip()
print(prompt.strip())

print(clean_prompt.upper())

print(clean_prompt.lower())

print(clean_prompt.find("awesome"))

print(clean_prompt.find("who"))

print(clean_prompt.count("w"))

print(clean_prompt.count("who"))

print(clean_prompt.count("Who"))

print(clean_prompt.replace("is", "are"))

print(clean_prompt.split())


Weather App Project i. taking weather data from real live data source ii. free API from OpenWeahterMap.org website
Fetch Weather Data Using OpenWeatherMap AP and create a dataset (Advanced - Optional) Objective: You will create a JSON file containing weather data for multiple cities over a period of 15 minutes. This task will help you practice your skills in making API calls, handling JSON responses, and converting data into a DataFrame.

Cities to Include:

New York, USA London, UK Tokyo, Japan Sydney, Australia Paris, France Instructions:

Obtain Your API Key: Go to the OpenWeatherMap website. Create a free account if you don’t have one. Navigate to the API section and generate your API key. Write the Python Code: Create a Python script that performs the following: Import the necessary libraries (requests, json, pandas). Define the list of cities you want to query and your API key. Set up a loop to make requests to the OpenWeatherMap API every 1 minute for 15 minutes. For each iteration, parse the JSON response and extract the following information: City name Temperature (in Celsius) Weather description Humidity Wind speed Timestamp Store this information in a list of dictionaries. Convert the list of dictionaries to a DataFrame. Save the DataFrame to a JSON file. Deliverable: Store this information in a list of dictionaries. Convert the list of dictionaries to a DataFrame. Save the DataFrame to a JSON file." You must in the end have a dataset with 15 entries of wheather for each city

API connected Gradio App i. API for quotes, jokes & advice to Gradio ii. App to generate text from APIs
random_quote_url = "https://api.quotable.io/random"

Fetches a random quote with author and content
random_joke_url = "https://official-joke-api.appspot.com/random_joke"

Fetches a random joke with setup and punchline
random_advice_url = "https://api.adviceslip.com/advice"

Fetches random life advice slip with advice text



The Scenario 1:
 
- The school's database has just exported a batch of raw exam scores. You need to use NumPy to generate some sample data, inspect its structure, and automatically grade the students based on a curve.
 
Your Tasks:
 
1. The Setup: Import NumPy. To ensure the whole class gets the exact same "random" numbers, set the random seed to 42.
 
2. Data Generation: Use random.randint to generate a 2D array of random scores between 0 and 100. The array should represent 10 students and their scores across 3 different subjects.
 
3. Array Inspection: Prove you built the array correctly by printing its ndim (number of dimensions), its shape, and its total size (total number of scores).
 
4. Student IDs: Use arange to create a 1D array of Student IDs starting at 100 and ending at 110.
 
5. Perform the aggregate operation of average and find the average score of students for each subject
 
6. The Audit (Boolean Masking): The principal wants to know every instance where a student scored below a 50. Use boolean masking to extract and print just those failing scores from the main array.
 
7. Automated Grading (where): Use np.where to create a brand new "Report Card" array. If a score in the original array is less than 50, it should say "Fail". If it is 50 or above, it should say "Pass".

In [ ]:
import numpy as np

# student_ids = np.arange(100, 110)

# student_names = np.array(["Ali",
                          "Abu",
                          "Ahmad",
                          "Zaki",
                          "Osman",
                          "Muthu",
                          "Kumar",
                          "Loh Min Nah",
                          "Johnny",
                          "Sarah",
])


# scores = np.array([[62,68,72],
                  [46,80,72],
                  [64,70,62],
                  [56,82,62],
                  [90,88,92],
                  [62,88,70],
                  [40,50,78],
                  [50,30,38],
                  [28,42,70],
                  [40,70,26]
])

In [ ]:
# Task 1: Import NumPy and set random seed
print(np.random.seed(42))

# Task 2: Generate a 2D array of random scores
# 10 students, 3 subjects, scores between 0 and 100
scores = np.random.randint(0, 101, (10, 3))

print("Exam Scores: ")
print(scores)

# Task 3: Array Inspection
print("\nArray Inspection")
print("Number of Dimensions (ndim):", scores.ndim)
print("Shape:", scores.shape)
print("Total Size:", scores.size)

# Task 4: Create Student IDs from 100 to 109
student_ids = np.arange(100, 110)

print("\nStudent IDs:")
print(student_ids)



None
Exam Scores:
[[51 92 14]
 [71 60 20]
 [82 86 74]
 [74 87 99]
 [23  2 21]
 [52  1 87]
 [29 37  1]
 [63 59 20]
 [32 75 57]
 [21 88 48]]

Array Inspection
Number of Dimensions (ndim): 2
Shape: (10, 3)
Total Size: 30

Student IDs:
[100 101 102 103 104 105 106 107 108 109]


In [8]:
# Task 5: Average score for each subject
subject_average = np.mean(scores, axis=0)

print("\nAverage Score for Each Subject:")
print(subject_average)

# Task 6: Boolean Masking
# Extract scores below 50

failing_scores = scores[scores < 50]

print("\nFailing Scores (Below 50):")
print(failing_scores)

# Task 7: Automated Grading using np.where

report_card = np.where(scores < 50, "Fail", "Pass")

print("\nReport Card:")
print(report_card)





Average Score for Each Subject:
[49.8 58.7 44.1]

Failing Scores (Below 50):
[14 20 23  2 21  1 29 37  1 20 32 21 48]

Report Card:
[['Pass' 'Pass' 'Fail']
 ['Pass' 'Pass' 'Fail']
 ['Pass' 'Pass' 'Pass']
 ['Pass' 'Pass' 'Pass']
 ['Fail' 'Fail' 'Fail']
 ['Pass' 'Fail' 'Pass']
 ['Fail' 'Fail' 'Fail']
 ['Pass' 'Pass' 'Fail']
 ['Fail' 'Pass' 'Pass']
 ['Fail' 'Pass' 'Fail']]


Scenario 2:
 
The semester is over. The principal wants a final report that includes Student IDs, their final scores, and a text label indicating whether they Passed or Failed. You need to combine all this data into one final grid and export it so it can be sent to the administration.
 
Your Tasks:
 
1. The Numerical Data: Access the 1D marks arrays of 5 Student IDs (101 to 105), store separately in variables .
 
2. The Text Data: Use `np.where` to create a status array. If the score is 50 or above, the status is "Pass". If it's below 50, it is "Fail".
 
3. The Merge (column_stack): The principal wants this data side-by-side. Use np.column_stack to merge the IDs, the Scores, and the Statuses into one single 2D array called final_report. Print it to the screen.
 
4. Data Types: The administration suddenly wants you to add 5 bonus points to the scores in your final_report. Try to add 5 to the score column. What happens? Print the .dtype (data type) of your final_report to figure out why.
 
(OPTIONAL)
 
5. The Export: Even though the math broke, save your final_report to your computer as a CSV file called "edupi_final_grades.csv" so another program can deal with it.
 

In [9]:
# Task 1:

import numpy as np

# Student IDs
student_ids = np.arange(101, 106)

# Final Scores
scores = np.array([85, 42, 67, 49, 91])

print("Student IDs:")
print(student_ids)

print("\nScores:")
print(scores)

# Task 2

status = np.where(scores >= 50, "Pass", "Fail")

print("Status:")
print(status)


# Task 3

final_report = np.column_stack((student_ids, scores, status))

print("Final Report:")
print(final_report)

Student IDs:
[101 102 103 104 105]

Scores:
[85 42 67 49 91]
Status:
['Pass' 'Fail' 'Pass' 'Fail' 'Pass']
Final Report:
[['101' '85' 'Pass']
 ['102' '42' 'Fail']
 ['103' '67' 'Pass']
 ['104' '49' 'Fail']
 ['105' '91' 'Pass']]


In [10]:
# task 4

score_column = final_report[:, 1].astype(int)

bonus_scores = score_column + 5

print(bonus_scores)


#task 5

np.savetxt(
    "edupi_final_grades.csv",
    final_report,
    delimiter=",",
    fmt="%s",
    header="StudentID,Score,Status",
    comments=""
)

print("CSV file saved successfully!")


[90 47 72 54 96]
CSV file saved successfully!
